sycl/source/detail/scheduler/graph_processor.cpp

```diff
@@ -87,7 +90,24 @@ bool Scheduler::GraphProcessor::enqueueCommand(Command *Cmd,
         return false;
   }
 
-  return Cmd->enqueue(EnqueueResult, Blocking);
+  {
+    // Only graph read lock is to be held here.
+    // Enqueue process of a command may last quite a time. Having graph locked
+    // can introduce some thread starving (i.e. when the other thread attempts
+    // to acquire write lock and add a command to graph).
+    // Releasing read lock without other safety measures isn't an option here as
+    // the other thread could go into graph cleanup process (due to some event
+    // complete) and remove some dependencies from dependencies of the user of
+    // this command. An example: command A depends on commands B and C. This
+    // wants to enqueue A. Hence, it needs to enqueue B and C. So this thread
+    // gets into dependency list and starts enqueueing B right away. The other
+    // thread waits on completion of C and starts cleanup process. This thread
+    // is still in the middle of enqueue of B. The other thread modifies
+    // dependency list of A by removing C out of it. Iterators become invalid.
+    bool Result = Cmd->enqueue(EnqueueResult, Blocking);
+
+    return Result;
+  }
```

Comment:
Non-blocking comment: Do you really need this additional block with { and } ?
I mean, why not just place the comment section right before 'return Cmd->enqueue(EnqueueResult, Blocking);' ?

Reply:
Removed the block.